### ISSUE TITLE
[DR] Add support for NN training and acceleration for moment computations in PKPM

### ISSUE THREAD

**Issue by JonathanGorard** [ammarhakim/gkylzero#730]
This work has been prototyped within the `kann_integration_prototyping` branch: https://github.com/ammarhakim/gkylzero/tree/kann_integration_prototyping

The intention here is to add support within our input file structure for the continuous training of neural networks for the acceleration of moment computations within the PKPM app (with the hopes that an equivalent API can also be extended to Vlasov and Gyrokinetics). We use the highly minimal `kann` (https://github.com/attractivechaos/kann) C framework for performing autodiff, and for representing and training our neural networks. One near-term objective is to construct fully magnetized closures for use in future 10-moment simulations by training a neural network on PKPM simulation data to predict the heat fluxes `q_par` and `q_perp` (and their derivatives) based on the densities and pressures `rho`, `p_par` and `p_perp` (and their derivatives). Since we intend to train directly on the DG expansions themselves, the configuration space derivatives of all of the necessary quantities will already exist within the neural network, without needing to be computed within the 10-moment simulation itself. We also want to be able to test a neural network that has already been trained, to determine whether it is able to predict the PKPM moments given the existing simulation data.

The proposed additions to the PKPM input file structure are:

- `train_nn` / `trainNN`: should we be training a neural network on this simulation data? This will be `false` by default, to prevent the slowing down of simulations, but users should be encouraged to set it to `true` whenever resources permit. When set to `true`, a separate neural network will be trained on each particle species with a given frequency.
- `train_ab_initio` / `trainAbInitio`: should we construct and train a new neural network from scratch (ab initio), or should we load in an existing neural network and continue training on that? This will be `true` by default, since we cannot assume that a user has any saved neural networks on their machine, but in practice anyone using this framework will likely want to set it to `false` to ensure that their training efforts compound cumulatively with each simulation that they run.
- `nn_width` / `NNWidth`: for an ab initio training run, how wide should the neural network be (i.e. how many neurons should be contained in each layer)? This will be 256 by default, to provide enough degrees of freedom to learn all of the necessary DG coefficients.
- `nn_depth` / `NNDepth`: for an ab initio training run, how deep should the neural network be (i.e. how many hidden layers should there be)? This will be 5 by default, to provide sufficient depth to learn moderately complex relationships between the input and output data, without also being prohibitively expensive.
- `train_nn_file` / `trainNNFile`: for a non-ab initio training run, what file path should we expect to find the existing neural network at? This will have a default value based on the simulation name.
- `num_trains` / `numTrains`: how many times should we train throughout the course of the simulation? This will be `GKYL_MAX_INT` by default, which will train the neural network on the results from every time-step, which will slow the simulation down considerably but will also provide the highest quality training data for the network. Other times we may only want to train on a few frames per simulation, just to get a little bit of extra training data but without things being slowed down too much.
- `num_nn_writes` / `numNNWrites`: how many times should we output the current state of the neural network throughout the course of the simulation? This will be 1 by default (just like with `num_frames`), so that the state of the neural network is output only at the beginning and end of the simulation, although it could be set higher for a longer simulation run so that intermediate training states can be obtained.
- `input_moms` / `inputMoms`: the list of PKPM moments to consider as "input" (e.g. `rho`, `p_par` and `p_perp` for the training of heat flux closures, which is the default case), expressed as integers. Since we train directly on the DG expansions, the first (and potentially second) configuration space derivatives of these moments are also included in the training data.
- `output_moms` / `outputMoms`: the list of PKPM moments to consider as "output" (e.g. `q_par` and `q_perp` for the training of heat flux closures, which is the default case), expressed as integers. Since we train directly on the DG expansions, the first (and potentially second) configuration space derivatives of these moments are also included in the training data.
- `test_nn` / `testNN`: should we be testing a neural network on this simulation data? This will be `false` by default, since we cannot assume that a user has any saved neural networks on their machine, but if they do then we can output the neural network's predicted values of the PKPM moments given the simulation data at any time-step. This will be saved to a file with the same structure as the standard PKPM diagnostic moments output, but with the suffix `..._moms_predicted` rather than simply `..._moms`.
- `test_nn_file` / `testNNFile`: for a testing run, what file path should we expect to find the neural network at? This will have a default value based on the simulation name.
- `num_tests` / `numTests`: for a testing run, how many times should we test (i.e. how frequently should we try to predict the values of the PKPM moments given the present simulation data) throughout the course of the simulation? This will be 1 by default, predicting only the initial and final diagnostic moments. These predictions will be saved to a file with the same structure as the standards PKPM moments output, but with the suffix `..._moms_predicted` rather than simply `..._moms`.

We use certain standard neural network hyperparameters for the training (i.e. learning rate 0.0001, batch size 64, maximum epochs 50, maximum dropout streak 10, validation data proportion 0.1, activation function hyperbolic tangent) that are not settable directly from the input file, to prevent overcrowding/overcomplication, since we don't want/expect most users to be tweaking these hyperparameters themselves.

An example of how the API may be used within a real PKPM simulation file:

![Image](https://github.com/user-attachments/assets/ff65ed93-492a-4f22-9483-61f5bd6b7b87)

And, just for fun, an example of a predicted profile for `q_par` (as predicted by a NN that has never seen this particular simulation data before):

![Image](https://github.com/user-attachments/assets/52aa4f91-6ee0-4c96-ad58-91087a6d32e0)
